platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/EdgeState.java

```diff
@@ -374,6 +374,7 @@ public Long getSelectedVariant() {
         if (node == null) {
             return null;
         } else {
+            assert node.getComponent() == getSelectedComponent();
```

Comment:
When the resolution engine is screwed up and starts returning corrupt graphs.

This is one of the conditions that our tests fixed. 

We are putting this into make sure it remains valid.

Reply:
Can you add a comment please?  Just a "// Should never fail".  Whenever I read an assert I think there's a possibility it could.

Reply:
Well... it could. If the code is broken. 

With that reasoning, all asserts would be marked with `// Should never fail`. You are asserting an assumption that you have during development, but can't necessarily ensure via the type system. 

If we add this comment here, we need to add it everywhere else we use assert

Reply:
I agree, and think you should describe every assert like this.  Many languages (especially newer ones like [Kotlin](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/assert.html)) give you an "argument" or something as part of the assert statement/method to document them in exactly this way.

Think about it like this - when the assert fails after a new change you make elsewhere next year, don't you want to quickly understand what assumption you just violated?

Reply:
But `Should never fail` does not give any more context to someone seeing this fail

Reply:
And you can add messages to Java asserts with assert condition: "message"

Reply:
Oh, ok, I forgot about that.  See, even older languages allow it!  So let's do it here!

And if you're saying "should never fail" isn't descriptive enough, then certainly "should never fail because of X" is even better.

Reply:
I'm going to leave this as-is